### PR TITLE
Make fusion handle pad/pack/unpack ops correctly.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This PR sets up the dispatch region formation for handling pad/pack fusion approriately to

- Fuse `tensor.pack` and `iree_linalg_ext.set_encoding` operations with its producers
- Fuse `tensor.unpack`/`iree_linalg_ext.unset_encoding` operations with either `tensor.extract_slice` or `linalg.generic` operations that are elementwise ops. This fusion is purposely limited since unpack -> consumer fusion is harder to generate good code for. Ideally we should not have `unpack` operations at all.
- `tensor.pad` operations generated during set encoding passes get pulled into the dispatch corresponding to `iree_linalg_ext.set_encoding`. This is a bit wonky right now suggesting need for an operation that is analogous to `iree_linalg_ext.set_encoding` but has padding semantics to make the lgoic here simpler.
- Adds a flag to control fusion of pad with producers. Currently IREE fuses pad with its consumers under a flag. Ideally the preference would be to fuse pad with its producers. This requires more work in backends to get things vectorized and code generated more effectively. To enable this work, a flag `iree-flow-fuse-padding-into-producer-ops` is added that allows a path for backends to start adding this support.
- There is also a flag `-iree-flow-enable-pad-handling` to disable the split of pad into `fill` -> `insert_slice`. This would be great to be on by default, but not all backends support this.

Summary for use of the pad related flags
- `iree-flow-enable-pad-handling` avoids splitting `pad` into `fill` -> `insert_slice`. It does not enable fusion of pad with producers or consumers.
- `iree-flow-enable-fuse-padding-into-linalg-producer-ops` avoids the split of `pad` into `fill` -> `insert_slice` and enables fusion of pad with producers.
- `iree-flow-enable-fuse-padding-into-linalg-consumer-ops` avoids the split of `pad` into `fill` -> `insert_slice` for pad ops that can be fused into its consumers. For now this only fuses pad op with consumers that implement the convolution op interface.

In the final state, all three of the above flags should be dropped, but that requires some work. These flags allow making progress towards this on all backends selectively. They are meant to be developer flags, and not meant to be user-facing supported flags.

### CODE REVIEW

**Review by hanhanW — APPROVED** [review]

**Comment by pzread** [comment]
Right now we are still using the  legacy benchmark suite to benchmark the Android phones. So in addition to the Python script, we still need to update `https://github.com/openxla/iree/blob/main/benchmarks/TFLite/android-arm64-v8a.cmake#L203`

**Comment by MaheshRavishankar** [comment]
> Right now we are still using the legacy benchmark suite to benchmark the Android phones. So in addition to the Python script, we still need to update `https://github.com/openxla/iree/blob/main/benchmarks/TFLite/android-arm64-v8a.cmake#L203`

Sigh... that sounds complicated... let me see if I can actually fix the problem.

**Comment by MaheshRavishankar** [comment]
@hanhanW I split the vectorize pack/unpack pass into two as we discussed. I think that does address my issue.. PTAL

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 1]
```diff
@@ -500,8 +502,6 @@ void addMultiTilingExpertPassPipeline(OpPassManager &passManager,
     // TODO(hanchung): Merge two vectorization passes into a pass. All the ops
     // should be vectorized altogether. Otherwise, there would be tensor.empty
     // ops which becomes a stack allocation in bufferization.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:506]: This TODO can be removed because they are vectorized altogether now. :)

**Review by hanhanW — APPROVED** [review]
LG, thanks!

**Review by hanhanW — COMMENTED** [review]
We probably have to fix a lit test here: https://github.com/openxla/iree/blob/4e2c85d69e9f1f3e42cfe8c07c2c1dcd49ff6363/compiler/src/iree/compiler/Codegen/LLVMCPU/test/data_tiling_pipeline.mlir#L75-L86

**Comment by MaheshRavishankar** [comment]
> We probably have to fix a lit test here:
> 
> https://github.com/openxla/iree/blob/4e2c85d69e9f1f3e42cfe8c07c2c1dcd49ff6363/compiler/src/iree/compiler/Codegen/LLVMCPU/test/data_tiling_pipeline.mlir#L75-L86

Doesnt seem to show up. It doesnt fail locally, nor seems to show up on CI. the split should be an NFC for the lit test.

**Comment by hanhanW** [comment]
> Doesnt seem to show up. It doesnt fail locally, nor seems to show up on CI. the split should be an NFC for the lit test.

That's surprising to me.. I'll take a look at it after you land the PR. The PR itself looks good to me.

**Comment by MaheshRavishankar** [comment]
@pzread or @hanhanW can you please advise on the failure here. I am just trying to add `"--iree-llvmcpu-fail-on-out-of-bounds-stack-allocation=false"` to the experimental flags for MobileNetV2 and MobileNetV3 as a workaround for a failure that this change is exposing.... What am I missing?

**Comment by hanhanW** [comment]
I skimmed through the python file and `.cmake` files, they look good to me. Not sure why the flag is not added to `iree-benchmark-android`. @pzread can you help take a look at it?

**Comment by pzread** [comment]
> I skimmed through the python file and `.cmake` files, they look good to me. Not sure why the flag is not added to `iree-benchmark-android`. @pzread can you help take a look at it?

I can't make suggested changes on these places
https://github.com/openxla/iree/pull/12744/files#diff-c5aae5c553ee3fed9e8ba2b9275cc9d03768d64c80a20f5370c6031c97231773R203
https://github.com/openxla/iree/pull/12744/files#diff-c5aae5c553ee3fed9e8ba2b9275cc9d03768d64c80a20f5370c6031c97231773R274

But they also need to be updated. There are 3 groups of benchmarks: sync, 1-thread, 4-thread. Right now we only updated the 4-thread.

**Comment by MaheshRavishankar** [comment]
> > I skimmed through the python file and `.cmake` files, they look good to me. Not sure why the flag is not added to `iree-benchmark-android`. @pzread can you help take a look at it?
> 
> I can't make suggested changes on these places [#12744 (files)](https://github.com/openxla/iree/pull/12744/files#diff-c5aae5c553ee3fed9e8ba2b9275cc9d03768d64c80a20f5370c6031c97231773R203) [#12744 (files)](https://github.com/openxla/iree/pull/12744/files#diff-c5aae5c553ee3fed9e8ba2b9275cc9d03768d64c80a20f5370c6031c97231773R274)
> 
> But they also need to be updated. There are 3 groups of benchmarks: sync, 1-thread, 4-thread. Right now we only updated the 4-thread.

Thanks! updated those. Hopefully that goes through.

**Pull request merged**